www/js/lib/uiUtil.js

```diff
@@ -172,6 +172,157 @@ function slideAway (e) {
+            obj.tagName = element.tagName;
+            headings.push(obj);
+        }
+        return headings;
+    };
+}
+
+/*
+    * Setup table of contents and display the list when the dropup button is clicked
+    * @param {List} a list of all headings as objects
+    * @returns {void}
+*/
+// Inject table of contents list into dropup element and scroll selection into view
+function setUpTOC () {
+    // var iframe = document.getElementById('articleContent');
+    var innerDoc = articleContainer.contentDocument;
+    var tableOfContents = new HeadingsTOC(innerDoc);
+    var headings = tableOfContents.getHeadingObjects();
+
+    var dropupHtml = '';
+    headings.forEach(function (heading) {
+        if (/^h1$/i.test(heading.tagName)) {
+            dropupHtml += '<li style="font-size:' + 18 + 'px;"><a style="color: black;" href="#" data-heading-id="' + heading.id + '">' + heading.textContent + '</a></li>';
+        } else if (/^h2$/i.test(heading.tagName)) {
+            dropupHtml += '<li style="margin-top:6px;margin-left:6px;font-size:' + 16 + 'px;"><a style="color: black;" href="#" data-heading-id="' + heading.id + '">' + heading.textContent + '</a></li>';
+        } else if (/^h3$/i.test(heading.tagName)) {
+            dropupHtml += '<li style="margin-left:12px;font-weight:350;font-size:' + 14 + 'px;"><a style="color: black;" href="#" data-heading-id="' + heading.id + '">' + heading.textContent + '</a></li>';
+        } else if (/^h4$/i.test(heading.tagName)) {
+            dropupHtml += '<li style="margin-left:16px;font-weight:300;font-size:' + 12 + 'px;"><a style="color: black;" href="#" data-heading-id="' + heading.id + '">' + heading.textContent + '</a></li>';
+        }
+        // Skip smaller headings (if there are any) to avoid making list too long
+    });
+    var ToCList = document.getElementById('ToCList');
+    ToCList.style.maxHeight = ~~(window.innerHeight * 0.75) + 'px';
+    ToCList.innerHTML = dropupHtml;
+    Array.from(ToCList.getElementsByTagName('a'))
+        .forEach(function (listElement) {
+            listElement.addEventListener('click', function () {
+                var sectionEle = innerDoc.getElementById(this.dataset.headingId)
+
+                var sectionsToOpen = getParentSections(sectionEle); // get all parents which are 'section' or 'details'
+                openSection(sectionsToOpen); // open all parents
+                // why..? because if the section is inside a details element, it will be closed by default
+
+                sectionEle.scrollIntoView();
+
+                // highlighting the section
+                sectionEle.style.backgroundColor = '#bdd1e5';
+                setTimeout(function () {
+                    sectionEle.style.backgroundColor = '';
+                }, 400);
+                sectionEle.style.transition = 'background-color 300ms ease-out';
+            });
+        });
+}
+
+document.addEventListener('DOMContentLoaded', function () {
```

Comment:
Please add a comment above to say what this code block does (I know what it does, but it's a good idea to comment explicitly and label it as part of the ToC setup code!